### PR TITLE
MPFR patch to fix sin on 32-bit long.

### LINKED ISSUE #22758: Loss of accuracy of BigFloat on Win10-64bit plattform
I found an issue regarding BigFloats using julia0.6.0 on a Win10 system. This error does not appear on linux or macOS systems:
```
# Set precision to 7 million bits, so I expect about 2 million valid digits
setprecision(7000000)
# Get value of pi with high precision
bigpi=BigFloat(pi)
# Calculate difference of sin(pi/2)-1. Should be zero in exact arithmetic
sin(bigpi/BigFloat(2))-1
# Calualate difference of sin(pi/6)-1/2. Should be zero in exact arithmetic
sin(bigpi/BigFloat(6))-1/BigFloat(2)
```

For the first difference I get the value of zero. That means the result of sin(pi/2) is valid up to 2 million digits.

For the seconds difference I get an value of about 4.76993...e-306923. That means sin(pi/6) is only calculated up to about 300,000 digits and not 2 million digits.

Some more details have already been discussed:
https://discourse.julialang.org/t/accuracy-of-bigfloat-calculating-sin-with-high-accuracy/4604

### PR DESCRIPTION
**Pull request by simonbyrne**
Fixes #22758.

### CODE REVIEW

**Comment by tkelman** [comment]
Cool. Can we modify the version info with some distinguishing marker that we could read through the C API, so we could add a test that only runs when MPFR is new enough or patched?

**Comment by simonbyrne** [comment]
Hmm, not that i know of. The other issue is that it actually takes a fair bit of time and memory to actually do the computation (especially in a VM), so I'm not sure it's that easy to test.

**Comment by tkelman** [comment]
How long are we talking, multiple minutes? I guess we could put it under `JULIA_TESTFULL` which generally only runs locally when manually asked for

**Comment by simonbyrne** [comment]
Actually, I think I found a way to do it. The bug shows up at lower precision, and the resulting test takes about 2 seconds on my computer, or 10 seconds in a 32-bit VM.

File: deps/patches/mpfr-sincos.patch [hunk 1]
```diff
@@ -17,3 +17,7 @@ svn diff -r 11589:11590 https://scm.gforge.inria.fr/anonscm/svn/mpfr/trunk
        /* the next term of the series is divided by Q[k] and multiplied
           by pp^2/2^(2r), thus the mult. factor < 1/2^mult[k] */
        MPFR_MPZ_SIZEINBASE2(mult[k], Q[k]);
+--- PATCHES	2017-07-18 15:55:44.000000000 -0700
++++ PATCHES	2017-07-18 15:55:57.000000000 -0700
```
Comment (tkelman) [inline deps/patches/mpfr-sincos.patch]: whoa neat

Comment (simonbyrne) [inline deps/patches/mpfr-sincos.patch]: yeah, it's not really well documented, but a nifty idea. I wish more projects did something like this.

**Comment by tkelman** [comment]
this lgtm but I should test it in a local windows source build to make sure we don't get any post merge surprises when this makes it into nightlies. will do so in the morning and merge if clear

File: base/mpfr.jl [hunk 2]
```diff
@@ -26,6 +26,9 @@ import Base.Math.lgamma_r
 
 import Base.FastMath.sincos_fast
 
+const MPFR_VERSION = VersionNumber(unsafe_string(ccall((:mpfr_get_version,:libmpfr), Ptr{Cchar}, ())))
+const MPFR_PATCHES = split(unsafe_string(ccall((:mpfr_get_patches,:libmpfr), Ptr{Cchar}, ())),' ')
```
Comment (tkelman) [inline base/mpfr.jl]: actually I do be wonder whether these should be set at runtime instead of sysimg build time

Comment (simonbyrne) [inline base/mpfr.jl]: I was just copying what GMP does. But yes, maybe we should.

The other issue is naming: LibGit2 simply names it `VERSION` (i.e. `LibGit2.VERSION`). Should we use this for consistency?

Comment (tkelman) [inline base/mpfr.jl]: or we could move this ccall to run in the tests until we have a more important need for it?

Comment (tkelman) [inline base/mpfr.jl]: GMP needed some consistency checks between sysimg build time libgmp version vs runtime libgmp version iirc, I don't know if old enough MPFR versions are still in common enough use for it to matter as much?

**Comment by simonbyrne** [comment]
I like the idea of exposing version information of libraries where possible.

File: base/mpfr.jl (hunk 2 above)
Comment (simonbyrne) [inline base/mpfr.jl]: Okay, I've changed it to a function for now. We should think about how to define a consistent interface for library versions (both for Base and in packages).

Comment (tkelman) [inline base/mpfr.jl]: agreed, and build-time vs run-time since they may differ (much more commonly for base than for packages, but never know)

**Review by tkelman — APPROVED** [review]
tested locally on both win32 and win64, passed

**Pull request merged**